### DISCUSSION TITLE
How to pause/resume main chare while workers are working?

### DISCUSSION THREAD

**Discussion by fclingen** [UIUC-PPL/charm#3377] [Q&A]
After some helpful pointing to the ldbRedDemo.f90 example for my last question, I still have some questions about how to add a Charm++ reduction to the legacy Fortran code I'm trying to refactor.

What I have, is 3,000 lines of code with some REALLY expensive computations on a number of multidimensional arrays in the middle of it, and a lot of serial code bracketing the expensive operations that I don't want distributed to multiple workers

The legacy code I have is all one big file.  a semi-pseudo-code is given below:

```
PROGRAM MAINPROGRAM
      COMPLEX A(NY,NZ,NX), B(NY,NZ,NX)! Every array in legacy code is modified in-place
      CALL LOAD_INIT_CONDITIONS_FROM_FILE (A,B,C,D) ! pull data from disk and do pre-processing to initialize dataset
      CALL PROCESS_A(A) ! A will be replaced with the processed version of A after calling
      CALL PROCESS_B(B) ! B will be replaced with the processed version of A after calling
      CALL WRITE_RESULTS(A,B)
END PROGRAM

SUBROUTINE PROCESS_A(A)
COMPLEX A (NY,NZ,NX)  ! Note the index order--none of the "layers" in the code are actually in contiguous memory since Fortran is column-major
REAL U(NZ,NX),V(NZ,NX),W(NZ,NX) !there are actually 10-50 of these arrays in the real code
REAL L(NX,NX), M(NZ,NX), N(NZ,NX)
REAL INPUT_PARAMS(NZ,NX)
INTEGER I

DO I=1,NY  ! Every NY layer DOES NOT depend on results from other layers.  In practice NY is roughly  50-ish
      CALL SETUP_UVW(I, A,U,V,W) ! Computations actually happen on U, V, and W, and the results of those computations get inserted into A one (NZ,NX) layer at a time
      
      CALL FFT_THINGS(U) !  These subroutines are where the legacy code spends 80% of its processing time
      CALL FFT_THINGS(V)  !  Each call to PROCESS is mutually independent, though the call to AGGREGATE needs results from all the work
      CALL FFT_THINGS(W) ! This is what I'm wanting to put into a chare array

! THIS IS WHERE I WOULD NEED TO SUSPEND A SUB-CHARE AND WAIT FOR RESULTS

      CALL DO_SOME_ARITHMETIC(U,V,W,L,M,N) ! This function is cheap, but needs results from all of U,V,W to run

     CALL MORE_FFT_THINGS(L)
     CALL MORE_FFT_THINGS(M)
     CALL MORE_FFT_THINGS(N)

     CALL AGGREGATE_UVWLMN(I,A,U,V,W,L,M,N)  ! The I-th (NZ,NX) layer will be modified here
END DO

! THIS IS WHERE I WOULD NEED TO SUSPEND THE MAIN CHARE TO WAIT FOR RESULTS

     CALL WRITE_TO_AN_OUTPUT_FILE(A)
END SUBROUTINE
```
IDEALLY, I want every call to FFT_THINGS or MORE_FFT_THINGS to happpen in its own chare, nested withing every NY layer happening within it's own chare (so I could have up to 300-ish calls to FFT_THINGS happening concurrently).  In my mind, refactoring this code means refactoring the PROGRAM into a subroutine, and calling that subroutine within a main chare.  After that, all the refactoring is going to happen in PROCESS_A and PROCESS_B--for both, it means Invoking entry methods on chars that are responsible for inidividual layers, which in turn invoke chares responsible for doing the invidiual FFTs.

As far as I've gotten, however, is just trying to get a main chare that will invoke the FFT functions and properly get back results using the Fortran90 bindings.  What has me stuck is figuring out exactly how to get results out of chares called from the main chare, for a number of reasons:

1.   I don't know how to get main chare to asynchronously launch chares, suspend while the chares work, then resume where it left off once they have returned results.  I think I have enough of a grok on Charm++ I understand how it should work in principle with CkCallbacks, but I don't know how those are implemented in f90.  The only example I've seen in Fortran callbacks is with reductions (Big Showstopper).  
2. Speaking of reductions, I don't see any way to pass back one piece of the larger array back at a time using reductions.  Again, the mechanism in Charm++ seems to exist in CkReduction::concat, but that reduction type is not one of the "simple" reduction types available in f90charm.  It appears if I want to do that type of reduction I'd have to do some janky stuff with a SUM-type reduction on an array that's zeroed out everywhere but in the area where the workers did their work (janky, but not a show-stopper)
3. There's going to be A LOT of copying data to re-shape it (because processing is happening on strided data) once I actually get to the point of parallelizing the loop (still kinda janky, but not so bad)

**Comment by rbuch** [comment]
I'm not sure I fully understand your question, but hopefully this helps a bit. So long as the main chare has no pending messages for it, it will "wait" in an idle state until it receives something. The canonical way to do the sort of synchronization you're talking about is through reductions (perhaps multiple reductions if you need to synchronize over chare arrays of multiple types). `AtSync` isn't truly a barrier, the call doesn't block and `ResumeFromSync` is called immediately if there is no "receiver" for `AtSync` ("receivers" are basically services that will be called when every participating element has called `AtSync`). The semantics of `AtSync` are that any object that calls `AtSync` will later get its `ResumeFromSync` called by the runtime, your chare won't resume from where it called `AtSync`.

**Reply by fclingen** [reply]
Part of my issue, is that I'm learning and doing software engineering at the same time.

I have Fortran code that has serial sections that have to stay serial, and large quantities of mutally-independent, computation heavy operation that I want to parallelize.  

In my own investigation since posting the question, I've come to understand that using reductions is canonically the way this is done (though in my case I really need a concat operation which I've discovered isn't implemented in f90charm, but I can muddle through with a sum instead) but I'm trying to work out how exactly I can construct my code so everything stays synchronized, and I don't essentially have to run the serial portion of the code I'm trying to modify on every chare.  

I THOUGHT that meant doing a reduction on my worker chares, then sending a message to the main chare once the reduction is finished (and thus the main chare needs a way to sit and wait until the workers have results to send), but I guess what's actually needed is multiple reductions?  I thought reductions were a "within array" operation, how do you perform a reduction across multiple chare types?

**Reply by rbuch** [reply]
Yes, reductions are a "within array" operation, I only mentioned multiple reductions since I wasn't sure if you had multiple arrays (perhaps of multiple types), they're only necessary if that's the case. Doing a reduction and then sending a message to the main chare is the right thing to do in general.

**Reply by fclingen** [reply]
How do you make the main chare block execution until the message is sent and resume operation once it has the results?  I don't see a way to set up semaphores or other similar synchronization constructs, it seems like they're all baked into the contribute and/or atsync functions, which is why I thought I needed to somehow finagle atsync do my blocking for me?

**Reply by rbuch** [reply]
Execution in Charm++ is message driven and asynchronous by default, so you shouldn't need to explicitly block. Objects will implicitly wait until they receive a message, so you can just send a message to an entry method in the main chare via a callback in the reduction, for example.

**Reply by fclingen-pet** [reply]
Looking in the docs, it seems like  a CkThreadedCallback is the object I would need for the behavior I desire--IF I were using C++. 

I haven't seen anything in the Fortran90 examples where callbacks are used outside of reductions.   How could I implement a threaded callback in f90?

**Reply by sanjaykale** [reply]
Maybe an orthogonal point to the above discussion: I hope you are not trying to pause the constructor method of the main chare. The threaded method should be some other method of the main chare.

**Reply by sanjaykale** [reply]
Some thoughts.. but I am writing somewhat blind because I am only guessing what you are trying to do (so please excuse the irrelevant bit):  First, based on your description, you should not need threaded methods, and CkThreadedCallback etc. at all. The method that receives reduction results can be in the main chare and continues whatever it wanted to do after getting the results. Also, AtSync is not very useful *unless* you are repeating this operation : fire tasks to workers, get results back, do something with it , for many iterations.  3rd: I hope it is clear that the workers will not have access to the whole array, if that array is either a global variable or inside the main chare.  You have to partition the array so that each chare has a portion of it. (unless you want to explicity pass the whole array to everyone, OR it is readonly data that you want all workers to have access to.  (It seems clear that we need more examples for f90Charm. Maybe you can contribute some by writing a simple variant of what you are trying to do). Also: concat as a reduction type is available. (Although I caution: you better have computational work in each chare much larger than the cost of transferring the data back and forth between main chare and worker chares.

**Reply by fclingen** [reply]
I updated my OP to maybe make it more clear what I'm trying to accomplish.

There is a LOT of computation happening within each chare (in principal), but I don't think concat is a reduction type available?   When I look here:
https://github.com/UIUC-PPL/charm/blob/main/src/ck-core/charmmod.f90
I don't see a parameter definition for a concat-type callback.  I'd assumed that meant it hadn't been implemented fro f90?

**Reply by sanjaykale** [reply]
You are right. The C++ version does support concat. https://charm.readthedocs.io/en/latest/charm++/manual.html?highlight=concat#built-in-reduction-types but based on what you show, fortran version does not. But this might be an omission easy to fix. OTOH, if the data each chare is contributing is large, it may be ok to send it in separate direct message to the root. At least as a stopgap solution.

**Reply by fclingen** [reply]
That still leaves me uncertain if how to actually synchronize passing the message to the root.  I'm back to wondering how to implement a CkCallbackReserveThread in Fortran

**Reply by fclingen** [reply]
OK, after playing with it some it seems I can do a concat-type reduction--I modified the ldbRedDemo to make it happen.  It's just not represented in the parameters listed in charmmod.f90.  However, if you put '68' in for the reduction-type argument in <chare>_contribute() the system knows what to do with it.

**Reply by fclingen** [reply]
I think I'm getting somewhere.  It took a bit of creativity, but I got something workable for my application.

Instead of having multiple chare classes, I have a single 1D chare array with a a threaded "main" entry method that runs the code I am overhauling, as well as entry methods that do the parallelizable bits.  I run the main entry method on a single chare, and when I want to parallelize computation I invoke the parallelizable methods from my main chare and immediately call CkWaitQD to let them finish their computations and message passing.  When they're done CkWaitQD will resume my serial code where it left off.

**Comment by stwhite91** [comment]
Yes, this sounds like a good approach. We can work on adding real support for concat and other types of reductions to the Fortran interface too.

**Reply by fclingen** [reply]
While I haven't tested it fully (as in, not doing any instrumentation or anything, just running with what works), it seems like it'd be pretty easy, just add a `CHARM_CONCAT=68` parameter to charmmod.f90

In order to use `CkWaitQD`, I had add `include "ckwaitdqd"` and add an `extern "C" void ckwaitqd_(void){CkWaitQD()}` to my .C file so Fortran could see it.  That would probably be better in the code that generates the '.decl.h' file.
